In [1]:
import pandas as pd
import numpy as np

#### Basic Dataframe Examination

Start off with a basic examination of the dataframes using `Pandas` standard EDA functions.

In [15]:
species = pd.read_csv('data/species_info.csv')
observations = pd.read_csv('data/observations.csv')

In [3]:
species.head(3)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN

In [4]:
observations.head(3)

scientific_name                            park_name  observations
0  Vicia benghalensis  Great Smoky Mountains National Park            68
1      Neovison vison  Great Smoky Mountains National Park            77
2   Prunus subcordata               Yosemite National Park           138

In [5]:
species.shape, observations.shape

((5824, 4), (23296, 3))

In [6]:
species.info(), print("\n"), observations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


(None, None, None)

Regular Size for the notebook

<font size="4">There are only null values in the species dataframe are under the column conservation status. Some more familiarization of the species data set would be helpful before making decision on the null values.</font>

<span style="font-size: 20px;"> test </span>

<p>More Testing</p>

#### heading 4 size


In [7]:
# species.conservation_status.value_counts()
species.conservation_status.value_counts(dropna=False)

NaN                   5633
Species of Concern     161
Endangered              16
Threatened              10
In Recovery              4
Name: conservation_status, dtype: int64

In [8]:
species.category.value_counts()

Vascular Plant       4470
Bird                  521
Nonvascular Plant     333
Mammal                214
Fish                  127
Amphibian              80
Reptile                79
Name: category, dtype: int64

In [9]:
species.scientific_name.nunique(), observations.scientific_name.nunique(), species.common_names.nunique()

(5541, 5541, 5504)

<font size="3">The dataframe observations only contains the scientific name and not the common name. There are 37 less total number of common names compared to scientific names, indicating that some species may share a common name. There are 5824 total species in the data frame indicating 283 possible duplicates in the species scientific names.<br><br>The observations data frame also contains 5541 unique scientific names. To make sure all species are included in both, a check is performed to see if the unique names from both data sets are equal.</font>

In [10]:
# Casting the list to a set could also be used here
sorted(observations.scientific_name.unique()) == sorted(species.scientific_name.unique())

True

<font size="3">Considering the fact that the vast majority of species have **NaN** for conservation status, and no species scientific names are missing in the observation data, it is safe to say that since all the animals in the species data are being observed, that a missing conservation status means that the species is not under any threat.<br><br>Thus, the null values will be changed to a new category *"Not in Danger"* after some other analyis.<br><br> Null Values will be treated as a species not under conservation watch.</font>

In [12]:
# species.conservation_status.isna().sum()

In [13]:
# species.fillna('Not in Danger', inplace=True)
# species

In [21]:
species.conservation_status.value_counts().sum()

191

From looking at the value count, we know there are only 191 species that are _Under Watch_ which I will define as having some dangerous conservation status.

Let's look at the percentage of each species under watch based on their Class (or category in the data)

In [22]:
cat_counts =species.category.value_counts().to_dict()

In [23]:
cat_counts =species.category.value_counts().to_dict()
under_watch = species.groupby(['category', 'conservation_status'])['scientific_name'].count().reset_index()
def percentages(row):
    return row['scientific_name']/cat_counts[row['category']]
under_watch['percentage'] = under_watch.apply(percentages, axis=1)
under_watch.sort_values(by='percentage', ascending=False)

category conservation_status  scientific_name  percentage
5                Bird  Species of Concern               72    0.138196
11             Mammal  Species of Concern               28    0.130841
14            Reptile  Species of Concern                5    0.063291
1           Amphibian  Species of Concern                4    0.050000
9              Mammal          Endangered                7    0.032710
7                Fish  Species of Concern                4    0.031496
8                Fish          Threatened                4    0.031496
2           Amphibian          Threatened                2    0.025000
6                Fish          Endangered                3    0.023622
13  Nonvascular Plant  Species of Concern                5    0.015015
0           Amphibian          Endangered                1    0.012500
16     Vascular Plant  Species of Concern               43    0.009620
12             Mammal          Threatened                2    0.009346
3                Bird          Endangered                4    0.007678
4                Bird         In Recovery                3    0.005758
10             Mammal         In Recovery                1    0.004673
17     Vascular Plant          Threatened                2    0.000447
15     Vascular Plant          Endangered                1    0.000224

It is clear from the chart that most species are not in danger. Especially with plants where only around 1% of species are of concern or worse. Species from the animal kingdom, on the other hand, have around 10% or higher of the species under conservation watch.

Let's move on to investigate duplicate entries.

In [24]:
species.duplicated().sum()

0

There are no complete duplicate rows indicating that any duplication is coming from the scientific name or common name only, as per our counts earlier. 

I suspect there are different common names for some species and some scientific names duplicated. 

Let's take a look at how many there are of each.

In [26]:
species.duplicated(subset=['scientific_name']).sum(), species.duplicated(subset=['common_names']).sum()

(283, 320)

Let's see which share both scientific and common names

In [36]:
species[species.duplicated(subset=['scientific_name','common_names'], keep=False)].sort_values(by='common_names')

category      scientific_name     common_names conservation_status
3020   Mammal          Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal          Canis lupus  Gray Wolf, Wolf          Endangered
560      Fish  Oncorhynchus mykiss    Rainbow Trout                 NaN
3283     Fish  Oncorhynchus mykiss    Rainbow Trout          Threatened

## NOTE HERE ABOUT THESE TWO
Only two species have shared scientific and common names. In addition, they also have different status. 
Let's look at species with duplicated common names first to see their conservation status.

In [27]:
# pd.set_option('display.max_rows', 30)
duplicated_common_names = species[species.duplicated(subset=['common_names'], keep=False)]
duplicated_common_names.sort_values(by='common_names')#[~duplicated_common_names.conservation_status.isna()]

category                      scientific_name  \
2730  Nonvascular Plant              Dichodontium pellucidum   
2822  Nonvascular Plant               Cirriphyllum piliferum   
2022     Vascular Plant                       Carex normalis   
1971     Vascular Plant                      Carex bromoides   
1960     Vascular Plant                      Carex annectens   
...                 ...                                  ...   
250                Bird                   Dendroica coronata   
252                Bird                   Dendroica dominica   
3206               Bird                   Setophaga dominica   
2957  Nonvascular Plant  Zygodon viridissimus var. rupestris   
2956  Nonvascular Plant                 Zygodon viridissimus   

                 common_names conservation_status  
2730                   A Moss                 NaN  
2822                   A Moss                 NaN  
2022           A Sedge, Sedge                 NaN  
1971           A Sedge, Sedge                 NaN  
1960           A Sedge, Sedge                 NaN  
...                       ...                 ...  
250     Yellow-Rumped Warbler                 NaN  
252   Yellow-Throated Warbler                 NaN  
3206  Yellow-Throated Warbler                 NaN  
2957             Zygodon Moss                 NaN  
2956             Zygodon Moss                 NaN  

[568 rows x 4 columns]

In [30]:
duplicated_common_names.conservation_status.value_counts(dropna=False)

NaN                   556
Species of Concern      9
In Recovery             1
Threatened              1
Endangered              1
Name: conservation_status, dtype: int64

The majority of the duplicated species are not in danger. Below, I will create a list of those species who are under watch.

In [32]:
duplicated_common_names.sort_values(by='common_names')[~duplicated_common_names.conservation_status.isna()]

/Users/robertjfpal/opt/anaconda3/envs/the_way/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


category                      scientific_name  \
2929  Nonvascular Plant                  Bazzania nudicaulis   
185                Bird                     Guiraca caerulea   
2292     Vascular Plant                       Poa paludigena   
3102               Bird                    Stellula calliope   
3020             Mammal                          Canis lupus   
4448             Mammal                          Canis lupus   
3288               Fish                       Cottus bairdii   
284                Bird                Vermivora ruficapilla   
310                Bird                     Contopus cooperi   
3283               Fish                  Oncorhynchus mykiss   
4879     Vascular Plant  Plagiobothrys torreyi var. diffusus   
287                Bird                     Wilsonia pusilla   

                common_names conservation_status  
2929                Bazzania  Species of Concern  
185            Blue Grosbeak  Species of Concern  
2292           Bog Bluegrass  Species of Concern  
3102    Calliope Hummingbird  Species of Concern  
3020         Gray Wolf, Wolf         In Recovery  
4448         Gray Wolf, Wolf          Endangered  
3288         Mottled Sculpin  Species of Concern  
284        Nashville Warbler  Species of Concern  
310   Olive-Sided Flycatcher  Species of Concern  
3283           Rainbow Trout          Threatened  
4879  Torrey's Popcornflower  Species of Concern  
287         Wilson's Warbler  Species of Concern

In [33]:
mask = duplicated_common_names.conservation_status == 'Species of Concern'
duplicated_common_names[mask]

category                      scientific_name  \
185                Bird                     Guiraca caerulea   
284                Bird                Vermivora ruficapilla   
287                Bird                     Wilsonia pusilla   
310                Bird                     Contopus cooperi   
2292     Vascular Plant                       Poa paludigena   
2929  Nonvascular Plant                  Bazzania nudicaulis   
3102               Bird                    Stellula calliope   
3288               Fish                       Cottus bairdii   
4879     Vascular Plant  Plagiobothrys torreyi var. diffusus   

                common_names conservation_status  
185            Blue Grosbeak  Species of Concern  
284        Nashville Warbler  Species of Concern  
287         Wilson's Warbler  Species of Concern  
310   Olive-Sided Flycatcher  Species of Concern  
2292           Bog Bluegrass  Species of Concern  
2929                Bazzania  Species of Concern  
3102    Calliope Hummingbird  Species of Concern  
3288         Mottled Sculpin  Species of Concern  
4879  Torrey's Popcornflower  Species of Concern

In [37]:
spec_of_concern = duplicated_common_names[mask].common_names.to_list()
duplicated_common_names[duplicated_common_names.common_names.isin(spec_of_concern)].sort_values(by='common_names')

category                      scientific_name  \
2930  Nonvascular Plant                  Bazzania tricrenata   
2929  Nonvascular Plant                  Bazzania nudicaulis   
2928  Nonvascular Plant                    Bazzania denudata   
4525               Bird                   Passerina caerulea   
185                Bird                     Guiraca caerulea   
2292     Vascular Plant                       Poa paludigena   
5628     Vascular Plant         Poa leptocoma ssp. leptocoma   
3102               Bird                    Stellula calliope   
4506               Bird                 Selasphorus calliope   
563                Fish                        Cottus bairdi   
3288               Fish                       Cottus bairdii   
3199               Bird              Oreothlypis ruficapilla   
284                Bird                Vermivora ruficapilla   
309                Bird                    Contopus borealis   
310                Bird                     Contopus cooperi   
4880     Vascular Plant   Plagiobothrys torreyi var. torreyi   
4879     Vascular Plant  Plagiobothrys torreyi var. diffusus   
287                Bird                     Wilsonia pusilla   
3196               Bird                   Cardellina pusilla   

                common_names conservation_status  
2930                Bazzania                 NaN  
2929                Bazzania  Species of Concern  
2928                Bazzania                 NaN  
4525           Blue Grosbeak                 NaN  
185            Blue Grosbeak  Species of Concern  
2292           Bog Bluegrass  Species of Concern  
5628           Bog Bluegrass                 NaN  
3102    Calliope Hummingbird  Species of Concern  
4506    Calliope Hummingbird                 NaN  
563          Mottled Sculpin                 NaN  
3288         Mottled Sculpin  Species of Concern  
3199       Nashville Warbler                 NaN  
284        Nashville Warbler  Species of Concern  
309   Olive-Sided Flycatcher                 NaN  
310   Olive-Sided Flycatcher  Species of Concern  
4880  Torrey's Popcornflower                 NaN  
4879  Torrey's Popcornflower  Species of Concern  
287         Wilson's Warbler  Species of Concern  
3196        Wilson's Warbler                 NaN

This initial suspicion appears to be the case. Some of these animals or plants with different common names are variations on a species. For all extent and purposes, they are separate entities. Most of them are not in danger, while 12 of them have different converstation statuses. All will be kept as seperate entities since the scientific names are different.

Further investigation of the Grey Wolf and Rainbow Trout will be necessary.

Next, let's look at the duplicated scientific names.

In [38]:
duplicated_scientific_names = species[species.duplicated(subset=['scientific_name'], keep=False)].sort_values(by='scientific_name')
duplicated_scientific_names

category      scientific_name  \
5553  Vascular Plant  Agrostis capillaris   
2132  Vascular Plant  Agrostis capillaris   
2134  Vascular Plant    Agrostis gigantea   
5554  Vascular Plant    Agrostis gigantea   
4178  Vascular Plant   Agrostis mertensii   
...              ...                  ...   
5643  Vascular Plant        Vulpia myuros   
2331  Vascular Plant     Vulpia octoflora   
4290  Vascular Plant     Vulpia octoflora   
3347  Vascular Plant         Zizia aptera   
632   Vascular Plant         Zizia aptera   

                                           common_names conservation_status  
5553                  Colonial Bent, Colonial Bentgrass                 NaN  
2132                                  Rhode Island Bent                 NaN  
2134                                             Redtop                 NaN  
5554                Black Bent, Redtop, Water Bentgrass                 NaN  
4178               Arctic Bentgrass, Northern Bentgrass                 NaN  
...                                                 ...                 ...  
5643  Foxtail Fescue, Rattail Fescue, Rat-Tail Fescu...                 NaN  
2331                                      Annual Fescue                 NaN  
4290  Eight-Flower Six-Weeks Grass, Pullout Grass, S...                 NaN  
3347  Heartleaf Alexanders, Heart-Leaf Alexanders, M...                 NaN  
632                                   Golden Alexanders                 NaN  

[557 rows x 4 columns]

In [45]:
duplicated_scientific_names[['conservation_status']].value_counts(dropna=False)

conservation_status
NaN                    534
Species of Concern      19
Endangered               2
In Recovery              1
Threatened               1
dtype: int64

Some species are in the database twice because they have different common names. Like with common names, a majority of them are not in danger. It is possible that these should be considered as different entities since data collection might be occurring on different groups of these species, but we are not sure yet. 

Let's focus on the 23 species that are under watch.

In [47]:
print(duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count().sum())
duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count()

23


conservation_status  scientific_name          
Endangered           Canis lupus                  2
In Recovery          Canis lupus                  1
Species of Concern   Eptesicus fuscus             2
                     Gavia immer                  2
                     Lasionycteris noctivagans    2
                     Myotis californicus          2
                     Myotis lucifugus             3
                     Nycticorax nycticorax        2
                     Pandion haliaetus            2
                     Riparia riparia              2
                     Taxidea taxus                2
Threatened           Oncorhynchus mykiss          1
Name: common_names, dtype: int64

In [49]:
duplicated_scientific_names.scientific_name.value_counts().value_counts()

2    265
3      9
Name: scientific_name, dtype: int64

Only 9 species total have 3 scientific names. It appears that the Canis lupus (the grey wolf) and Myotis lucifugus (the little brown bat) both have three statuses under watch. We already know that about the wolf and we also know that the rainbow trout has _threatened_ and _not in danger_ statuses.

Let's look more at the species on this list to see if any of the entries with a count of 2 have a mixed statuses and confirm the statuses of species of three.

Myotis lucifugus         3
Puma concolor            3
Castor canadensis        3
Procyon lotor            3
Streptopelia decaocto    3
Canis lupus              3
Columba livia            3
Hypochaeris radicata     3
Holcus lanatus           3
Name: scientific_name, dtype: int64

In [51]:
list_of_3_entries = duplicated_scientific_names.scientific_name.value_counts()[duplicated_scientific_names.scientific_name.value_counts() == 3].index
list_of_2_entries = duplicated_scientific_names.scientific_name.value_counts()[duplicated_scientific_names.scientific_name.value_counts() == 2].index

In [52]:
duplicated_scientific_names[duplicated_scientific_names.scientific_name.isin(list_of_3_entries)]

category        scientific_name  \
8             Mammal            Canis lupus   
3020          Mammal            Canis lupus   
4448          Mammal            Canis lupus   
49            Mammal      Castor canadensis   
4475          Mammal      Castor canadensis   
3050          Mammal      Castor canadensis   
4513            Bird          Columba livia   
3138            Bird          Columba livia   
156             Bird          Columba livia   
4236  Vascular Plant         Holcus lanatus   
5600  Vascular Plant         Holcus lanatus   
2242  Vascular Plant         Holcus lanatus   
3488  Vascular Plant   Hypochaeris radicata   
4799  Vascular Plant   Hypochaeris radicata   
858   Vascular Plant   Hypochaeris radicata   
4467          Mammal       Myotis lucifugus   
3042          Mammal       Myotis lucifugus   
37            Mammal       Myotis lucifugus   
3030          Mammal          Procyon lotor   
26            Mammal          Procyon lotor   
4459          Mammal          Procyon lotor   
16            Mammal          Puma concolor   
4451          Mammal          Puma concolor   
3022          Mammal          Puma concolor   
4514            Bird  Streptopelia decaocto   
3140            Bird  Streptopelia decaocto   
3077            Bird  Streptopelia decaocto   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
4448                                    Gray Wolf, Wolf          Endangered  
49                              American Beaver, Beaver                 NaN  
4475                                             Beaver                 NaN  
3050                                    American Beaver                 NaN  
4513                                        Rock Pigeon                 NaN  
3138              Common Pigeon, Rock Dove, Rock Pigeon                 NaN  
156                                           Rock Dove                 NaN  
4236                  Common Velvetgrass, Yorkshire-Fog                 NaN  
5600     Common Velvetgrass, Velvetgrass, Yorkshire Fog                 NaN  
2242                   Common Velvet Grass, Velvetgrass                 NaN  
3488          Spotted Cats-Ear, Hairy Cats-Ear, Gosmore                 NaN  
4799  Common Cat's-Ear, False Dandelion, Frogbit, Go...                 NaN  
858                        Cat's Ear, Spotted Cat's-Ear                 NaN  
4467                                Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern  
3030          Common Raccoon, Northern Raccoon, Raccoon                 NaN  
26                              Common Raccoon, Raccoon                 NaN  
4459                                            Raccoon                 NaN  
16                              Panther (Mountain Lion)                 NaN  
4451                                      Mountain Lion                 NaN  
3022                        Cougar, Mountain Lion, Puma                 NaN  
4514                             Eurasian Collared Dove                 NaN  
3140     Eurasian Collared Dove, Eurasian Collared-Dove                 NaN  
3077                             Eurasian Collared-Dove                 NaN

Since the list of 2 is large (265), let's create a mask to filter out any species that have both statuses as _not in danger_. We will test the mask first with the list of 3.

In [53]:
# duplicated_scientific_names[~(duplicated_scientific_names.conservation_status == "Not in Danger") & duplicated_scientific_names.scientific_name.isin(list_of_3_entries)]
mask = ~(duplicated_scientific_names.conservation_status.isna()) & duplicated_scientific_names.scientific_name.isin(list_of_3_entries)
duplicated_scientific_names[mask]#.scientific_name.nunique()

category   scientific_name  \
8      Mammal       Canis lupus   
3020   Mammal       Canis lupus   
4448   Mammal       Canis lupus   
4467   Mammal  Myotis lucifugus   
3042   Mammal  Myotis lucifugus   
37     Mammal  Myotis lucifugus   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
4448                                    Gray Wolf, Wolf          Endangered  
4467                                Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern

In [81]:
# duplicated_scientific_names[duplicated_scientific_names.scientific_name.isin(list_of_3_entries)]

In [54]:
# duplicated_scientific_names[~(duplicated_scientific_names.conservation_status == "Not in Danger") & duplicated_scientific_names.scientific_name.isin(list_of_2_entries)]
mask = ~(duplicated_scientific_names.conservation_status.isna()) & duplicated_scientific_names.scientific_name.isin(list_of_2_entries)
duplicated_scientific_names[mask]
# list_of_2_entries.shape[0]
# (duplicated_scientific_names[duplicated_scientific_names.conservation_status.isna()].shape[0] + duplicated_scientific_names[mask].shape[0]) / 2

category            scientific_name  \
29     Mammal           Eptesicus fuscus   
3035   Mammal           Eptesicus fuscus   
3150     Bird                Gavia immer   
172      Bird                Gavia immer   
30     Mammal  Lasionycteris noctivagans   
3037   Mammal  Lasionycteris noctivagans   
4465   Mammal        Myotis californicus   
3039   Mammal        Myotis californicus   
337      Bird      Nycticorax nycticorax   
4564     Bird      Nycticorax nycticorax   
3283     Fish        Oncorhynchus mykiss   
3081     Bird          Pandion haliaetus   
104      Bird          Pandion haliaetus   
226      Bird            Riparia riparia   
3185     Bird            Riparia riparia   
3029   Mammal              Taxidea taxus   
4457   Mammal              Taxidea taxus   

                                           common_names conservation_status  
29                                        Big Brown Bat  Species of Concern  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
172                                         Common Loon  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
4465                                  California Myotis  Species of Concern  
3039  California Myotis, California Myotis, Californ...  Species of Concern  
337                           Black-Crowned Night-Heron  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern  
3283                                      Rainbow Trout          Threatened  
3081                             Osprey, Western Osprey  Species of Concern  
104                                              Osprey  Species of Concern  
226                                        Bank Swallow  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
3029                            American Badger, Badger  Species of Concern  
4457                                             Badger  Species of Concern

All duplicated species are either on the list 2 or 3 times. Almost all of the species with multiple entries have the same conservation status. The majority here are 'Not in Danger, and 9 species with a status of 'Species of Concern'. Only two species, the gray wolf and rainbow trout, have multiple statuses. 

This means that the duplicate entries are most likely not by mistake, but were inlcuded because of the change in status.

As for the others, it is crucial to look at the obervations table to get a sense of what is going with the multiple entries.

In [136]:
# duplicated_scientific_names_2 = species[species.duplicated(subset=['scientific_name'], keep='first')].sort_values(by='scientific_name')

In [ ]:
# duplicated_scientific_names[duplicated_scientific_names.conservation_status.isna()]

In [101]:
duplicated_mask = species.duplicated(subset=['scientific_name'], keep=False)
trout_keep_NA_mask = species.scientific_name == species.loc[560, 'scientific_name']
duplicated_mask_NA = species.conservation_status.isna() & ~trout_keep_NA_mask

 = species[~(duplicated_mask & duplicated_mask_NA)]
# species[species.conservation_status.isna()].drop_duplicates(subset=['scientific_name'], keep='first')

IndentationError: unexpected indent (<ipython-input-101-27c49a733251>, line 5)

In [ ]:
species.shape

In [55]:
duplicated_scientific_names = species[species.duplicated(subset=['scientific_name'], keep=False)].sort_values(by='scientific_name')
duplicated_scientific_names

category      scientific_name  \
5553  Vascular Plant  Agrostis capillaris   
2132  Vascular Plant  Agrostis capillaris   
2134  Vascular Plant    Agrostis gigantea   
5554  Vascular Plant    Agrostis gigantea   
4178  Vascular Plant   Agrostis mertensii   
...              ...                  ...   
5643  Vascular Plant        Vulpia myuros   
2331  Vascular Plant     Vulpia octoflora   
4290  Vascular Plant     Vulpia octoflora   
3347  Vascular Plant         Zizia aptera   
632   Vascular Plant         Zizia aptera   

                                           common_names conservation_status  
5553                  Colonial Bent, Colonial Bentgrass       Not in Danger  
2132                                  Rhode Island Bent       Not in Danger  
2134                                             Redtop       Not in Danger  
5554                Black Bent, Redtop, Water Bentgrass       Not in Danger  
4178               Arctic Bentgrass, Northern Bentgrass       Not in Danger  
...                                                 ...                 ...  
5643  Foxtail Fescue, Rattail Fescue, Rat-Tail Fescu...       Not in Danger  
2331                                      Annual Fescue       Not in Danger  
4290  Eight-Flower Six-Weeks Grass, Pullout Grass, S...       Not in Danger  
3347  Heartleaf Alexanders, Heart-Leaf Alexanders, M...       Not in Danger  
632                                   Golden Alexanders       Not in Danger  

[557 rows x 4 columns]

In [59]:
duplicated_scientific_names_mixed_status = duplicated_scientific_names[~duplicated_scientific_names.conservation_status.isna()]['scientific_name'].unique().tolist()

In [62]:
species[species.scientific_name.isin(duplicated_scientific_names_mixed_status)]

category            scientific_name  \
8      Mammal                Canis lupus   
29     Mammal           Eptesicus fuscus   
30     Mammal  Lasionycteris noctivagans   
37     Mammal           Myotis lucifugus   
104      Bird          Pandion haliaetus   
172      Bird                Gavia immer   
226      Bird            Riparia riparia   
337      Bird      Nycticorax nycticorax   
560      Fish        Oncorhynchus mykiss   
3020   Mammal                Canis lupus   
3029   Mammal              Taxidea taxus   
3035   Mammal           Eptesicus fuscus   
3037   Mammal  Lasionycteris noctivagans   
3039   Mammal        Myotis californicus   
3042   Mammal           Myotis lucifugus   
3081     Bird          Pandion haliaetus   
3150     Bird                Gavia immer   
3185     Bird            Riparia riparia   
3283     Fish        Oncorhynchus mykiss   
4448   Mammal                Canis lupus   
4457   Mammal              Taxidea taxus   
4465   Mammal        Myotis californicus   
4467   Mammal           Myotis lucifugus   
4564     Bird      Nycticorax nycticorax   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
29                                        Big Brown Bat  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern  
104                                              Osprey  Species of Concern  
172                                         Common Loon  Species of Concern  
226                                        Bank Swallow  Species of Concern  
337                           Black-Crowned Night-Heron  Species of Concern  
560                                       Rainbow Trout                 NaN  
3020                                    Gray Wolf, Wolf         In Recovery  
3029                            American Badger, Badger  Species of Concern  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
3039  California Myotis, California Myotis, Californ...  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
3081                             Osprey, Western Osprey  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
3283                                      Rainbow Trout          Threatened  
4448                                    Gray Wolf, Wolf          Endangered  
4457                                             Badger  Species of Concern  
4465                                  California Myotis  Species of Concern  
4467                                Little Brown Myotis  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern

In [105]:
species[species.scientific_name == 'Oncorhynchus mykiss']

category      scientific_name   common_names conservation_status
560      Fish  Oncorhynchus mykiss  Rainbow Trout                 NaN
3283     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened

In [106]:
species[~duplicated_mask].conservation_status.value_counts()

Species of Concern    142
Endangered             14
Threatened              9
In Recovery             3
Name: conservation_status, dtype: int64

In [107]:
# # pd.set_option('display.max_rows', 30)
duplicated_scientific_names[duplicated_scientific_names.scientific_name.apply(lambda species: species in duplicated_scientific_names_mixed_status)]

category            scientific_name  \
8      Mammal                Canis lupus   
3020   Mammal                Canis lupus   
4448   Mammal                Canis lupus   
29     Mammal           Eptesicus fuscus   
3035   Mammal           Eptesicus fuscus   
3150     Bird                Gavia immer   
172      Bird                Gavia immer   
30     Mammal  Lasionycteris noctivagans   
3037   Mammal  Lasionycteris noctivagans   
4465   Mammal        Myotis californicus   
3039   Mammal        Myotis californicus   
4467   Mammal           Myotis lucifugus   
3042   Mammal           Myotis lucifugus   
37     Mammal           Myotis lucifugus   
337      Bird      Nycticorax nycticorax   
4564     Bird      Nycticorax nycticorax   
3283     Fish        Oncorhynchus mykiss   
560      Fish        Oncorhynchus mykiss   
3081     Bird          Pandion haliaetus   
104      Bird          Pandion haliaetus   
226      Bird            Riparia riparia   
3185     Bird            Riparia riparia   
3029   Mammal              Taxidea taxus   
4457   Mammal              Taxidea taxus   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
4448                                    Gray Wolf, Wolf          Endangered  
29                                        Big Brown Bat  Species of Concern  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
172                                         Common Loon  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
4465                                  California Myotis  Species of Concern  
3039  California Myotis, California Myotis, Californ...  Species of Concern  
4467                                Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern  
337                           Black-Crowned Night-Heron  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern  
3283                                      Rainbow Trout          Threatened  
560                                       Rainbow Trout                 NaN  
3081                             Osprey, Western Osprey  Species of Concern  
104                                              Osprey  Species of Concern  
226                                        Bank Swallow  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
3029                            American Badger, Badger  Species of Concern  
4457                                             Badger  Species of Concern

In [ ]:
######### HERE

In [60]:
duplicated_scientific_names[duplicated_scientific_names.scientific_name.isin(obsv_counts[:9].index)].sort_values(by='scientific_name')

NameError: name 'obsv_counts' is not defined

In [61]:
observations[observations.scientific_name.isin(obsv_counts[:9].index)].sort_values(by=['scientific_name', 'park_name', 'observations']).head(20)

NameError: name 'obsv_counts' is not defined

In [ ]:
back_up = species.copy()

In [ ]:
species = back_up
species.sort_values(by=['scientific_name', 'common_names'], inplace=True)
species.reset_index(drop=True, inplace=True)

In [62]:
species[species.scientific_name == 'Canis lupus']

category scientific_name     common_names conservation_status
8      Mammal     Canis lupus        Gray Wolf          Endangered
3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

In [63]:
species_df = pd.concat([species, species, species, species], ignore_index=True)

In [64]:
species_df = species_df.sort_values(by=['scientific_name', 'common_names']).reset_index(drop=True)#.drop(columns='scientific_name')
# species_df.drop(columns='scientific_name', inplace=True)

In [65]:
species_df[species_df.scientific_name == 'Canis lupus']

category scientific_name     common_names conservation_status
3352   Mammal     Canis lupus        Gray Wolf          Endangered
3353   Mammal     Canis lupus        Gray Wolf          Endangered
3354   Mammal     Canis lupus        Gray Wolf          Endangered
3355   Mammal     Canis lupus        Gray Wolf          Endangered
3356   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
3357   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered
3358   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
3359   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered
3360   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
3361   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered
3362   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
3363   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

In [66]:
observations_df = observations.sort_values(by=['scientific_name']).reset_index(drop=True)

In [67]:
observations[observations.scientific_name == 'Canis lupus'].sort_values(by=['park_name', 'observations'])

scientific_name                            park_name  observations
1766      Canis lupus                  Bryce National Park            27
7346      Canis lupus                  Bryce National Park            29
9884      Canis lupus                  Bryce National Park            74
17756     Canis lupus  Great Smoky Mountains National Park            14
10190     Canis lupus  Great Smoky Mountains National Park            15
20353     Canis lupus  Great Smoky Mountains National Park            30
10268     Canis lupus            Yellowstone National Park            60
10907     Canis lupus            Yellowstone National Park            67
13427     Canis lupus            Yellowstone National Park           203
1294      Canis lupus               Yosemite National Park            35
19987     Canis lupus               Yosemite National Park            44
19330     Canis lupus               Yosemite National Park           117

In [68]:
species_df = species_df.drop(columns='scientific_name')
df = pd.concat([species_df, observations_df], axis=1)
df

category                                       common_names  \
0         Vascular Plant                          Rocky Mountain Alpine Fir   
1         Vascular Plant                          Rocky Mountain Alpine Fir   
2         Vascular Plant                          Rocky Mountain Alpine Fir   
3         Vascular Plant                          Rocky Mountain Alpine Fir   
4         Vascular Plant  Balsam Fir, Colorado Fir, Concolor Fir, Silver...   
...                  ...                                                ...   
23291  Nonvascular Plant                                       Zygodon Moss   
23292  Nonvascular Plant                                       Zygodon Moss   
23293  Nonvascular Plant                                       Zygodon Moss   
23294  Nonvascular Plant                                       Zygodon Moss   
23295  Nonvascular Plant                                       Zygodon Moss   

      conservation_status                      scientific_name  \
0           Not in Danger                        Abies bifolia   
1           Not in Danger                        Abies bifolia   
2           Not in Danger                        Abies bifolia   
3           Not in Danger                        Abies bifolia   
4           Not in Danger                       Abies concolor   
...                   ...                                  ...   
23291       Not in Danger                 Zygodon viridissimus   
23292       Not in Danger  Zygodon viridissimus var. rupestris   
23293       Not in Danger  Zygodon viridissimus var. rupestris   
23294       Not in Danger  Zygodon viridissimus var. rupestris   
23295       Not in Danger  Zygodon viridissimus var. rupestris   

                                 park_name  observations  
0                Yellowstone National Park           215  
1                      Bryce National Park           109  
2      Great Smoky Mountains National Park            72  
3                   Yosemite National Park           136  
4      Great Smoky Mountains National Park           101  
...                                    ...           ...  
23291  Great Smoky Mountains National Park            71  
23292                  Bryce National Park           102  
23293               Yosemite National Park           210  
23294            Yellowstone National Park           237  
23295  Great Smoky Mountains National Park           102  

[23296 rows x 6 columns]

In [69]:
df[df.scientific_name == 'Canis lupus'].sort_values(by=['common_names', 'conservation_status'])

category     common_names conservation_status scientific_name  \
3352   Mammal        Gray Wolf          Endangered     Canis lupus   
3353   Mammal        Gray Wolf          Endangered     Canis lupus   
3354   Mammal        Gray Wolf          Endangered     Canis lupus   
3355   Mammal        Gray Wolf          Endangered     Canis lupus   
3357   Mammal  Gray Wolf, Wolf          Endangered     Canis lupus   
3359   Mammal  Gray Wolf, Wolf          Endangered     Canis lupus   
3361   Mammal  Gray Wolf, Wolf          Endangered     Canis lupus   
3363   Mammal  Gray Wolf, Wolf          Endangered     Canis lupus   
3356   Mammal  Gray Wolf, Wolf         In Recovery     Canis lupus   
3358   Mammal  Gray Wolf, Wolf         In Recovery     Canis lupus   
3360   Mammal  Gray Wolf, Wolf         In Recovery     Canis lupus   
3362   Mammal  Gray Wolf, Wolf         In Recovery     Canis lupus   

                                park_name  observations  
3352  Great Smoky Mountains National Park            30  
3353               Yosemite National Park            35  
3354            Yellowstone National Park            60  
3355                  Bryce National Park            27  
3357               Yosemite National Park            44  
3359                  Bryce National Park            74  
3361  Great Smoky Mountains National Park            14  
3363                  Bryce National Park            29  
3356            Yellowstone National Park            67  
3358               Yosemite National Park           117  
3360  Great Smoky Mountains National Park            15  
3362            Yellowstone National Park           203

In [70]:
####### STOP HERE

In [71]:
observations.groupby(['park_name', 'scientific_name']).sum()

observations
park_name              scientific_name                                  
Bryce National Park    Abies bifolia                                 109
                       Abies concolor                                 83
                       Abies fraseri                                 109
                       Abietinella abietina                          101
                       Abronia ammophila                              92
...                                                                  ...
Yosemite National Park Zonotrichia leucophrys gambelii               169
                       Zonotrichia leucophrys oriantha               135
                       Zonotrichia querula                           160
                       Zygodon viridissimus                          159
                       Zygodon viridissimus var. rupestris           210

[22164 rows x 1 columns]

In [72]:
observations.columns

Index(['scientific_name', 'park_name', 'observations'], dtype='object')

In [73]:
obsv_counts = observations.scientific_name.value_counts()
obsv_counts

Myotis lucifugus                        12
Puma concolor                           12
Hypochaeris radicata                    12
Holcus lanatus                          12
Streptopelia decaocto                   12
                                        ..
Packera dimorphophylla var. paysonii     4
Smilax bona-nox                          4
Chondestes grammacus                     4
Leymus triticoides                       4
Dichanthelium depauperatum               4
Name: scientific_name, Length: 5541, dtype: int64

In [74]:
all(obsv_counts[obsv_counts.values > 4].index.sort_values() == duplicated_scientific_names.scientific_name.unique())

True

This implies that there are no duplicates in the dataset. Any species with a duplicate name are separate observations from different times.

In [75]:
# duplicated_scientific_names_mixed_status.sort()
# duplicated_scientific_names_mixed_status


In [76]:
obsv_counts[obsv_counts.values == 8]

Lactuca biennis           8
Digitaria ischaemum       8
Brassica rapa             8
Linaria vulgaris          8
Eragrostis cilianensis    8
                         ..
Lanius excubitor          8
Anthemis cotula           8
Bidens tripartita         8
Hieracium caespitosum     8
Polygonum convolvulus     8
Name: scientific_name, Length: 265, dtype: int64

This is also show that for each species, they are observed only in multiples of 4, meaning that a species is recorded at each park either 1, 2, or 3 times.

It appears that the Gray Wolf was endangered but recovered, while the Rainbow Trout were no longer under threat or of concern, at least in the 4 National Parks under study

In [77]:
observations[observations.scientific_name == 'Canis lupus'].sort_values(by='observations')

scientific_name                            park_name  observations
17756     Canis lupus  Great Smoky Mountains National Park            14
10190     Canis lupus  Great Smoky Mountains National Park            15
1766      Canis lupus                  Bryce National Park            27
7346      Canis lupus                  Bryce National Park            29
20353     Canis lupus  Great Smoky Mountains National Park            30
1294      Canis lupus               Yosemite National Park            35
19987     Canis lupus               Yosemite National Park            44
10268     Canis lupus            Yellowstone National Park            60
10907     Canis lupus            Yellowstone National Park            67
9884      Canis lupus                  Bryce National Park            74
19330     Canis lupus               Yosemite National Park           117
13427     Canis lupus            Yellowstone National Park           203

In [78]:
observations[observations.scientific_name == 'Oncorhynchus mykiss'].sort_values(by='observations')

scientific_name                            park_name  observations
15020  Oncorhynchus mykiss  Great Smoky Mountains National Park            39
925    Oncorhynchus mykiss                  Bryce National Park            59
15239  Oncorhynchus mykiss               Yosemite National Park            59
3354   Oncorhynchus mykiss  Great Smoky Mountains National Park            61
11893  Oncorhynchus mykiss                  Bryce National Park           105
167    Oncorhynchus mykiss               Yosemite National Park           118
4649   Oncorhynchus mykiss            Yellowstone National Park           119
8682   Oncorhynchus mykiss            Yellowstone National Park           253

In [79]:
trout_ob = observations[observations.scientific_name == 'Oncorhynchus mykiss'].sort_values(by='observations')
trout_sp = species[species.scientific_name == 'Oncorhynchus mykiss']

In [80]:
trout_sp

category      scientific_name   common_names conservation_status
560      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger
3283     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened

In [81]:
trout_ob

scientific_name                            park_name  observations
15020  Oncorhynchus mykiss  Great Smoky Mountains National Park            39
925    Oncorhynchus mykiss                  Bryce National Park            59
15239  Oncorhynchus mykiss               Yosemite National Park            59
3354   Oncorhynchus mykiss  Great Smoky Mountains National Park            61
11893  Oncorhynchus mykiss                  Bryce National Park           105
167    Oncorhynchus mykiss               Yosemite National Park           118
4649   Oncorhynchus mykiss            Yellowstone National Park           119
8682   Oncorhynchus mykiss            Yellowstone National Park           253

In [82]:
trout = pd.merge(trout_sp, trout_ob, how='cross')

In [83]:
trout[(trout.conservation_status.isna()) & (trout.observations > 200) | ~trout.conservation_status.isna()]


category    scientific_name_x   common_names conservation_status  \
0      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
1      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
2      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
3      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
4      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
5      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
6      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
7      Fish  Oncorhynchus mykiss  Rainbow Trout       Not in Danger   
8      Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
9      Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
10     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
11     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
12     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
13     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
14     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   
15     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened   

      scientific_name_y                            park_name  observations  
0   Oncorhynchus mykiss  Great Smoky Mountains National Park            39  
1   Oncorhynchus mykiss                  Bryce National Park            59  
2   Oncorhynchus mykiss               Yosemite National Park            59  
3   Oncorhynchus mykiss  Great Smoky Mountains National Park            61  
4   Oncorhynchus mykiss                  Bryce National Park           105  
5   Oncorhynchus mykiss               Yosemite National Park           118  
6   Oncorhynchus mykiss            Yellowstone National Park           119  
7   Oncorhynchus mykiss            Yellowstone National Park           253  
8   Oncorhynchus mykiss  Great Smoky Mountains National Park            39  
9   Oncorhynchus mykiss                  Bryce National Park            59  
10  Oncorhynchus mykiss               Yosemite National Park            59  
11  Oncorhynchus mykiss  Great Smoky Mountains National Park            61  
12  Oncorhynchus mykiss                  Bryce National Park           105  
13  Oncorhynchus mykiss               Yosemite National Park           118  
14  Oncorhynchus mykiss            Yellowstone National Park           119  
15  Oncorhynchus mykiss            Yellowstone National Park           253

In [84]:
df = duplicated_scientific_names.merge(observations, on='scientific_name')

In [85]:
df.drop_duplicates(subset=['scientific_name', 'observations']).head(20)

category      scientific_name                          common_names  \
0   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
1   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
2   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
3   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
4   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
5   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
6   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
7   Vascular Plant  Agrostis capillaris     Colonial Bent, Colonial Bentgrass   
16  Vascular Plant    Agrostis gigantea                                Redtop   
17  Vascular Plant    Agrostis gigantea                                Redtop   
18  Vascular Plant    Agrostis gigantea                                Redtop   
19  Vascular Plant    Agrostis gigantea                                Redtop   
20  Vascular Plant    Agrostis gigantea                                Redtop   
21  Vascular Plant    Agrostis gigantea                                Redtop   
22  Vascular Plant    Agrostis gigantea                                Redtop   
23  Vascular Plant    Agrostis gigantea                                Redtop   
32  Vascular Plant   Agrostis mertensii  Arctic Bentgrass, Northern Bentgrass   
33  Vascular Plant   Agrostis mertensii  Arctic Bentgrass, Northern Bentgrass   
34  Vascular Plant   Agrostis mertensii  Arctic Bentgrass, Northern Bentgrass   
35  Vascular Plant   Agrostis mertensii  Arctic Bentgrass, Northern Bentgrass   

   conservation_status                            park_name  observations  
0        Not in Danger  Great Smoky Mountains National Park            84  
1        Not in Danger                  Bryce National Park           103  
2        Not in Danger                  Bryce National Park           105  
3        Not in Danger            Yellowstone National Park           241  
4        Not in Danger               Yosemite National Park           182  
5        Not in Danger            Yellowstone National Park           267  
6        Not in Danger  Great Smoky Mountains National Park            97  
7        Not in Danger               Yosemite National Park           140  
16       Not in Danger            Yellowstone National Park           253  
17       Not in Danger            Yellowstone National Park           235  
18       Not in Danger                  Bryce National Park           104  
19       Not in Danger  Great Smoky Mountains National Park            93  
20       Not in Danger                  Bryce National Park           116  
21       Not in Danger  Great Smoky Mountains National Park            57  
22       Not in Danger               Yosemite National Park           148  
23       Not in Danger               Yosemite National Park           128  
32       Not in Danger               Yosemite National Park           135  
33       Not in Danger               Yosemite National Park           128  
34       Not in Danger  Great Smoky Mountains National Park            54  
35       Not in Danger            Yellowstone National Park           271

In [86]:
print(f'mode\t{observations.observations.mode()}')
print(observations.observations.agg(['max', 'min', 'median', 'mean']))

mode	0    84
dtype: int64
max       321.000000
min         9.000000
median    124.000000
mean      142.287904
Name: observations, dtype: float64


In [87]:
pd.set_option('display.max_rows', 30)
observations[observations.scientific_name.apply(lambda species: species in duplicated_scientific_names_mixed_status)].sort_values(by=['scientific_name', 'observations']).head(30)



scientific_name                            park_name  observations
792    Agrostis capillaris  Great Smoky Mountains National Park            84
17428  Agrostis capillaris  Great Smoky Mountains National Park            97
3993   Agrostis capillaris                  Bryce National Park           103
4864   Agrostis capillaris                  Bryce National Park           105
17735  Agrostis capillaris               Yosemite National Park           140
7750   Agrostis capillaris               Yosemite National Park           182
6166   Agrostis capillaris            Yellowstone National Park           241
10379  Agrostis capillaris            Yellowstone National Park           267
11602    Agrostis gigantea  Great Smoky Mountains National Park            57
7763     Agrostis gigantea  Great Smoky Mountains National Park            93
6824     Agrostis gigantea                  Bryce National Park           104
8676     Agrostis gigantea                  Bryce National Park           116
17535    Agrostis gigantea               Yosemite National Park           128
13907    Agrostis gigantea               Yosemite National Park           148
483      Agrostis gigantea            Yellowstone National Park           235
449      Agrostis gigantea            Yellowstone National Park           253
775     Agrostis mertensii  Great Smoky Mountains National Park            54
3199    Agrostis mertensii                  Bryce National Park            74
9721    Agrostis mertensii  Great Smoky Mountains National Park            87
18314   Agrostis mertensii                  Bryce National Park            88
490     Agrostis mertensii               Yosemite National Park           128
111     Agrostis mertensii               Yosemite National Park           135
6169    Agrostis mertensii            Yellowstone National Park           251
1740    Agrostis mertensii            Yellowstone National Park           271
12730      Agrostis scabra  Great Smoky Mountains National Park            48
17994      Agrostis scabra                  Bryce National Park            60
7935       Agrostis scabra  Great Smoky Mountains National Park            91
23259      Agrostis scabra                  Bryce National Park           118
16448      Agrostis scabra               Yosemite National Park           162
18036      Agrostis scabra               Yosemite National Park           170

Once a species is over 250 in observation, they are no longer in danger or or concern|

Threatened appears to be under 100 observations in 3 or more parks while Under Concern seems to be under 100 in 2 or more parks

In [88]:
# observations[observations.duplicated(subset=['scientific_name'], keep=False)].sort_values(by='scientific_name')

The names are equal in bddoth data sets, so there would be no species that are in one but not in another

Looking at duplicated rows
Figure out if the duplicates are errors or not

In [89]:
observations[observations.duplicated(keep=False)].sort_values(by='scientific_name')

scientific_name                            park_name  \
513             Arctium minus               Yosemite National Park   
10674           Arctium minus               Yosemite National Park   
4527   Botrychium virginianum            Yellowstone National Park   
20294  Botrychium virginianum            Yellowstone National Park   
19392       Cichorium intybus            Yellowstone National Park   
14142       Cichorium intybus            Yellowstone National Park   
7263   Echinochloa crus-galli  Great Smoky Mountains National Park   
1454   Echinochloa crus-galli  Great Smoky Mountains National Park   
1020     Eleocharis palustris  Great Smoky Mountains National Park   
12381    Eleocharis palustris  Great Smoky Mountains National Park   
1032      Hesperis matronalis                  Bryce National Park   
13372     Hesperis matronalis                  Bryce National Park   
4178     Monotropa hypopithys  Great Smoky Mountains National Park   
20820    Monotropa hypopithys  Great Smoky Mountains National Park   
10038          Plantago major  Great Smoky Mountains National Park   
365            Plantago major  Great Smoky Mountains National Park   
18141           Poa compressa  Great Smoky Mountains National Park   
13097           Poa compressa  Great Smoky Mountains National Park   
17378    Potentilla norvegica               Yosemite National Park   
10809    Potentilla norvegica               Yosemite National Park   
7219        Prunella vulgaris  Great Smoky Mountains National Park   
16741       Prunella vulgaris  Great Smoky Mountains National Park   
19575            Salix exigua               Yosemite National Park   
22698            Salix exigua               Yosemite National Park   
17577      Spilogale putorius            Yellowstone National Park   
627        Spilogale putorius            Yellowstone National Park   
8710    Streptopelia decaocto            Yellowstone National Park   
1635    Streptopelia decaocto            Yellowstone National Park   
21151     Trifolium campestre            Yellowstone National Park   
18435     Trifolium campestre            Yellowstone National Park   

       observations  
513             162  
10674           162  
4527            232  
20294           232  
19392           266  
14142           266  
7263             62  
1454             62  
1020             62  
12381            62  
1032            124  
13372           124  
4178             73  
20820            73  
10038            90  
365              90  
18141            80  
13097            80  
17378           148  
10809           148  
7219             75  
16741            75  
19575           186  
22698           186  
17577           256  
627             256  
8710            255  
1635            255  
21151           239  
18435           239

In [90]:
duplicated_species = observations[observations.duplicated()]['scientific_name']

In [91]:
# duplicated_species

In [92]:
sorted_species_park = observations.sort_values(by=['scientific_name', 'park_name'])
sorted_species_park.tail(50)

scientific_name  \
9204    Zigadenus venenosus var. venenosus   
17255   Zigadenus venenosus var. venenosus   
10913                         Zizia aptera   
13904                         Zizia aptera   
7753                          Zizia aptera   
...                                    ...   
5539                  Zygodon viridissimus   
12775  Zygodon viridissimus var. rupestris   
20040  Zygodon viridissimus var. rupestris   
6879   Zygodon viridissimus var. rupestris   
16237  Zygodon viridissimus var. rupestris   

                                 park_name  observations  
9204             Yellowstone National Park           243  
17255               Yosemite National Park           150  
10913                  Bryce National Park           105  
13904                  Bryce National Park           112  
7753   Great Smoky Mountains National Park           105  
...                                    ...           ...  
5539                Yosemite National Park           159  
12775                  Bryce National Park           102  
20040  Great Smoky Mountains National Park           102  
6879             Yellowstone National Park           237  
16237               Yosemite National Park           210  

[50 rows x 3 columns]

Filter the data to see if duplicated data is legitimate. From the above observation, each specicies has at least 4 sets of observations with one from each park. However, some have multiple sets of observations.

In [93]:
sorted_species_park[['scientific_name', 'park_name']].value_counts()

scientific_name                      park_name                          
Procyon lotor                        Yellowstone National Park              3
Hypochaeris radicata                 Great Smoky Mountains National Park    3
Castor canadensis                    Great Smoky Mountains National Park    3
                                     Yellowstone National Park              3
                                     Yosemite National Park                 3
                                                                           ..
Equisetum scirpoides                 Great Smoky Mountains National Park    1
                                     Bryce National Park                    1
Equisetum laevigatum                 Yosemite National Park                 1
                                     Yellowstone National Park              1
Zygodon viridissimus var. rupestris  Yosemite National Park                 1
Length: 22164, dtype: int64

This shows that species can be observed muliptle times in one park, assuming this either on different time frames or by different observers

So, let's check if the species on the duplicated list seem to be a species with multiple observations

In [94]:
# pd.set_option('display.max_rows', 30)
sorted_species_park[sorted_species_park.scientific_name.apply(lambda species: species in duplicated_species.tolist())]

scientific_name                            park_name  observations
2054         Arctium minus                  Bryce National Park            66
16394        Arctium minus                  Bryce National Park           142
18320        Arctium minus  Great Smoky Mountains National Park            76
19102        Arctium minus  Great Smoky Mountains National Park            63
4017         Arctium minus            Yellowstone National Park           234
...                    ...                                  ...           ...
22131  Trifolium campestre  Great Smoky Mountains National Park            72
18435  Trifolium campestre            Yellowstone National Park           239
21151  Trifolium campestre            Yellowstone National Park           239
6478   Trifolium campestre               Yosemite National Park           130
8090   Trifolium campestre               Yosemite National Park           105

[124 rows x 3 columns]

The data above demonstrates that the rows that are duplicates are not errors because each species that contained a duplicate, had multiple observations from each of the four parks. I can conclude that a duplicate is just a recorded obervation done at a different time or by a different observer with the same value.

In [95]:
observations

scientific_name  \
0                           Vicia benghalensis   
1                               Neovison vison   
2                            Prunus subcordata   
3                         Abutilon theophrasti   
4                     Githopsis specularioides   
...                                        ...   
23291                     Croton monanthogynus   
23292                 Otospermophilus beecheyi   
23293  Heterotheca sessiliflora ssp. echioides   
23294                     Dicranella rufescens   
23295                           Cucurbita pepo   

                                 park_name  observations  
0      Great Smoky Mountains National Park            68  
1      Great Smoky Mountains National Park            77  
2                   Yosemite National Park           138  
3                      Bryce National Park            84  
4      Great Smoky Mountains National Park            85  
...                                    ...           ...  
23291               Yosemite National Park           173  
23292                  Bryce National Park           130  
23293                  Bryce National Park           140  
23294               Yosemite National Park           171  
23295               Yosemite National Park           164  

[23296 rows x 3 columns]

In [96]:
multiple_species = species.scientific_name.value_counts()

In [97]:
multiple_species.where(multiple_species > 1, inplace=True)

In [98]:
multiple_species.dropna(inplace=True)

In [99]:
multiple_species.value_counts()

2.0    265
3.0      9
Name: scientific_name, dtype: int64

In [100]:
species_changed_status = multiple_species.index.tolist()

In [101]:
species_df = species[~species.scientific_name.isin(species_changed_status)]

In [102]:
species_df.shape

(5267, 4)

In [103]:
species_df.duplicated().sum()

0

In [104]:
(species_df.groupby('scientific_name').count()['conservation_status'] == 1).sum()

5267

In [105]:
# species_df[~(species_df.conservation_status == "Not in Danger")]
species_df.head()

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
7   Mammal                  Canis latrans   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole       Not in Danger  
1                              American Bison, Bison       Not in Danger  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...       Not in Danger  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)       Not in Danger  
7                                             Coyote  Species of Concern

In [106]:
observations.head()

scientific_name                            park_name  observations
0        Vicia benghalensis  Great Smoky Mountains National Park            68
1            Neovison vison  Great Smoky Mountains National Park            77
2         Prunus subcordata               Yosemite National Park           138
3      Abutilon theophrasti                  Bryce National Park            84
4  Githopsis specularioides  Great Smoky Mountains National Park            85

In [107]:
observations.shape

(23296, 3)

Data merging safely from here

In [108]:
right = species_df.merge(observations, on='scientific_name', how='right').sort_values(by=['scientific_name', 'observations']).reset_index(drop=True)

In [109]:
left = species_df.join(observations.set_index('scientific_name'), on='scientific_name', how='left').sort_values(by=['scientific_name', 'observations']).reset_index(drop=True)

In [110]:
right[right.scientific_name.isin(species_changed_status)]

category      scientific_name common_names conservation_status  \
416        NaN  Agrostis capillaris          NaN                 NaN   
417        NaN  Agrostis capillaris          NaN                 NaN   
418        NaN  Agrostis capillaris          NaN                 NaN   
419        NaN  Agrostis capillaris          NaN                 NaN   
420        NaN  Agrostis capillaris          NaN                 NaN   
...        ...                  ...          ...                 ...   
23251      NaN         Zizia aptera          NaN                 NaN   
23252      NaN         Zizia aptera          NaN                 NaN   
23253      NaN         Zizia aptera          NaN                 NaN   
23254      NaN         Zizia aptera          NaN                 NaN   
23255      NaN         Zizia aptera          NaN                 NaN   

                                 park_name  observations  
416    Great Smoky Mountains National Park            84  
417    Great Smoky Mountains National Park            97  
418                    Bryce National Park           103  
419                    Bryce National Park           105  
420                 Yosemite National Park           140  
...                                    ...           ...  
23251                  Bryce National Park           112  
23252               Yosemite National Park           123  
23253               Yosemite National Park           129  
23254            Yellowstone National Park           257  
23255            Yellowstone National Park           269  

[2228 rows x 6 columns]

In [111]:
len(species_changed_status)

274